beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs

```diff
@@ -0,0 +1,241 @@
+
+        Ok(RPCRateLimiter {
+            ping_rl,
+            metadata_rl,
+            status_rl,
+            goodbye_rl,
+            bbroots_rl,
+            bbrange_rl,
+            init_time: Instant::now(),
+        })
+    }
+}
+
+impl RPCRateLimiter {
+    pub fn allows<T: EthSpec>(
+        &mut self,
+        peer_id: &PeerId,
+        request: &RPCRequest<T>,
+    ) -> Result<(), RateLimitedErr> {
+        let time_since_start = self.init_time.elapsed();
+        let tokens = request.expected_responses().max(1);
+        match request.protocol() {
+            Protocol::Ping => self.ping_rl.allows(time_since_start, peer_id, tokens),
+            Protocol::Status => self.status_rl.allows(time_since_start, peer_id, tokens),
+            Protocol::MetaData => self.metadata_rl.allows(time_since_start, peer_id, tokens),
+            Protocol::Goodbye => self.goodbye_rl.allows(time_since_start, peer_id, tokens),
+            Protocol::BlocksByRange => self.bbrange_rl.allows(time_since_start, peer_id, tokens),
+            Protocol::BlocksByRoot => self.bbroots_rl.allows(time_since_start, peer_id, tokens),
+        }
+    }
+}
+
+pub struct RateLimiter {
+    tau: Nanosecs,
+    t: Nanosecs,
+    /// Time when the bucket will be full for each peer. TAT (theoretical arrival time) from GCRA.
+    tat_per_key: FnvHashMap<PeerId, Nanosecs>,
+}
+
+impl RateLimiter {
+    pub fn from_quota(quota: Quota) -> Result<Self, &'static str> {
+        if quota.max_tokens == 0 {
+            return Err("Max number of tokens should be positive");
+        }
+        if quota.replenish_all_every.as_nanos() == 0 {
+            return Err("Replenish time must be positive");
+        }
+        let tau = quota
+            .replenish_all_every
+            .as_nanos()
+            .try_into()
+            .map_err(|_| "total replenish time is too long")?;
+        let t = (quota.replenish_all_every.as_nanos() / quota.max_tokens as u128)
+            .try_into()
+            .map_err(|_| "total replenish time is too long")?;
+        Ok(RateLimiter {
+            tau,
+            t,
+            tat_per_key: FnvHashMap::default(),
+        })
+    }
+
+    pub fn allows(
+        &mut self,
+        time_since_start: Duration,
+        peer_id: &PeerId,
+        tokens: u64,
+    ) -> Result<(), RateLimitedErr> {
+        let time_since_start = time_since_start.as_nanos() as u64;
+        let tau = self.tau;
+        let t = self.t;
+        // how long does it take to replenish these tokens, when at least one will be taken
+        let additional_time = t * (tokens - 1);
+        if additional_time + t > tau {
+            // the time required to process this amount of tokens is longer than the time that
+            // makes the bucket full. So, this batch can _never_ be processed
+            return Err(RateLimitedErr::TooLarge);
+        }
+        // If the peer is new, we consider their bucket full
```

Comment:
When new peers connect they immediately request probably Status, Ping and Metadata. 
Are we going to instantly reject these with a rate limit?